### DISCUSSION TITLE
Prompt output lost when using dynamic prompt variables, adetailer and batch count > 1

### DISCUSSION THREAD

**Discussion by zymox** [hako-mikan/sd-webui-regional-prompter#229] [Announcements]
This weirdly only seems to happen for me when I have both after detailer / adetailer and regional prompter active.
I'm not sure if it's an issue with regional prompter, dynamic prompts or after detailer.

**Example prompt:**

```
${nationality=__nationality__}${name=__girl_name__}
A ${nationality} woman named ${name} standing in a park BREAK 
A 21 year old woman, short hair, glasses BREAK 
A big tree, detailed wood
```

- Regional prompter active, split horizontal 1,1, attention, use base prompt.
- After detailer active, any prompt or now prompt, doesn't matter. 
- Batch count >= 2.

It will generate everything fine and seems to use the right dynamic prompts, but the meta data written to the PNG and shown in Automatic1111 repeat the prompt from the first image for every image after the first batch, instead of showing the new dynamic prompt for every image.

It works when I turn either regional prompter or adetailer off.

**Comment by zymox** [comment]
This happens due to you "restoring" p.all_prompts in process_batch() in rp.py.

In this case you make a copy all_prompts from a p which is itself a shallow, later discarded copy of p that adetailer makes. Adetailer later restores p.all_prompts themselves, but at this point, you have made your copy (self.all_prompts) from a wrong all_prompts already. Then you "restore" it to the actual used p in process_batch() when a new batch starts, and from then on it's messed up.

I'm not exactly sure why it is done like this, can't you maybe restore p.all_prompts in postprocess()?

**Comment by hako-mikan** [comment]
This is a necessary action to fill in the correct information to png info. 
It is possible to provide an option to disable this.
